### PR TITLE
Buttons parameter to array

### PR DESCRIPTION
**Pull request by DavidePastore**
This will change the type of `options.buttons` parameter as array. This will help us to choose the order of the buttons. Now you can set the `key` of the button using the `key` property.

Before:

``` javascript
bootbox.dialog({
  message: "I am a custom dialog",
  title: "Custom title",
  buttons: {
    success: {
      label: "Success!",
      className: "btn-success",
      callback: function() {
        Example.show("great success");
      }
    },
    danger: {
      label: "Danger!",
      className: "btn-danger",
      callback: function() {
        Example.show("uh oh, look out!");
      }
    },
    main: {
      label: "Click ME!",
      className: "btn-primary",
      callback: function() {
        Example.show("Primary button");
      }
    }
  }
});
```

With this PR:

``` javascript
bootbox.dialog({
  message: "I am a custom dialog",
  title: "Custom title",
  buttons: [
    {
      key: "success",
      label: "Success!",
      className: "btn-success",
      callback: function() {
        Example.show("great success");
      }
    },
    {
      key: "danger",
      label: "Danger!",
      className: "btn-danger",
      callback: function() {
        Example.show("uh oh, look out!");
      }
    },
    {
      key: "main",
      label: "Click ME!",
      className: "btn-primary",
      callback: function() {
        Example.show("Primary button");
      }
    }
  ]
});
```

Related issues: #291, #293, #431 and #525.

Things to do:
- [ ] WIP (#535) Update documentation 
- [ ] Create a migration guide

### CODE REVIEW

File: bootbox.js [hunk 1]
```diff
@@ -216,6 +216,24 @@
    * merge a set of default dialog options with user supplied arguments
    */
   function mergeArguments(defaults, args, properties) {
+    var mapArgumentsResult = mapArguments(
+      args,
+      properties
+    );
+    if(defaults && defaults.buttons && mapArgumentsResult && mapArgumentsResult.buttons){
```
Comment (tarlepp) [inline bootbox.js]: code style is different than anywhere else.

**Comment by tarlepp** [comment]
Also there should be some migration guide to convert old code for this. Or maybe just some serious notes is enough - although also docs needs to update.

File: bootbox.js (hunk 1 above)
Comment (DavidePastore) [inline bootbox.js]: Thanks for reporting the issue. Using the `grunt` command it's all fine (apart from the `getKeyLength` function). What are you referring to when you talk about code style?

Comment (tarlepp) [inline bootbox.js]: you're using eg. ```if(...){``` and everywhere else bootbox uses ```if (...) {```

Comment (DavidePastore) [inline bootbox.js]: Oh sorry, you're right! Is there a way to auto fix the code, so I use the same standard. Maybe a linter or something like this. I suppose it should be an automatic (`grunt`?) task.

File: bootbox.js [hunk 2]
```diff
@@ -258,38 +273,42 @@
...
-
+      buttons.push({
+        label: _t(value),
+        key: key
+      });
+      /*
       buttons[key] = {
         label: _t(value)
       };
+       */
```
Comment (tarlepp) [inline bootbox.js]: dead code?

Comment (DavidePastore) [inline bootbox.js]: You're right. :smile:

File: bootbox.js [hunk 3]
```diff
@@ -302,16 +307,22 @@
...
       if ($.isFunction(options.callback)) {
         return options.callback.call(this);
       }
       return true;
     };
 
+    /**
+     * overrides
+     */
+    each(options.buttons, function(buttonKey, button){
```
Comment (tarlepp) [inline bootbox.js]: another code style issue here use ```function(...) {```

File: bootbox.js [hunk 4]
```diff
@@ -216,6 +207,24 @@
    * merge a set of default dialog options with user supplied arguments
    */
   function mergeArguments(defaults, args, properties) {
+    var mapArgumentsResult = mapArguments(
+      args,
+      properties
+    );
+    if (defaults && defaults.buttons && mapArgumentsResult && mapArgumentsResult.buttons) {
```
Comment (tarlepp) [inline bootbox.js]: personally I like that code is readable, so I would add empty lines on these.

```
);
if (...)
```

to
```
);

if (...)
```

Comment (DavidePastore) [inline bootbox.js]: I agree with you. Is there any other place where you find the same issue?

Comment (tarlepp) [inline bootbox.js]: Didn't look closer, you should spot those quite easily.

File: bootbox.js [hunk 5]
```diff
@@ -216,6 +207,25 @@
...
   function mergeArguments(defaults, args, properties) {
+    var mapArgumentsResult = mapArguments(
+      args,
+      properties
+    );
+
+    if (defaults && defaults.buttons && mapArgumentsResult && mapArgumentsResult.buttons) {
+      $.each(defaults.buttons, function(defaultButtonKey, defaultButton) {
+        var found = false;
+        $.each(mapArgumentsResult.buttons, function(mapArgumentButtonKey, mapArgumentButton) {
```
Comment (tarlepp) [inline bootbox.js]: empty line for more readable code.

File: bootbox.js [hunk 6]
```diff
@@ -216,6 +207,25 @@
...
+          if (defaultButton.key === mapArgumentButton.key) {
+            found = true;
+          }
+        });
+        if (!found) {
```
Comment (tarlepp) [inline bootbox.js]: empty line for more readable code.

**Comment by DavidePastore** [comment]
@tarlepp Sorry If I misunderstand the **magic** line. :cry:  Do you think the review is ok?

**Comment by makeusabrew** [comment]
See my comment in #291. Need to remember why this changed, but it changed for a good reason.

**Pull request closed without merging**